### PR TITLE
Add Extract Component Function and Tests

### PR DESCRIPTION
**Pull request by Aristoeu**
For issue #332 Problem with ExtractComponent

### CODE REVIEW

File: python/mspasspy/algorithms/basic.py [hunk 1]
```diff
@@ -1,5 +1,55 @@
...
+    object_history=False,
+    alg_name=None,
+    alg_id=None,
+    dryrun=False,
+    inplace_return=False,
+    function_return_key=None,
+    **kwargs
+):
+    if isinstance(data, Seismogram):
+        empty = TimeSeries()
```
Comment (wangyinz) [inline python/mspasspy/algorithms/basic.py:28]: It is not a good idea to create an empty data object here. We only need it when the input is dead or exception occurs. Otherwise, this is only wasting CPU cycles and memory.

Also, the returned data object (whether being empty or not) should inherit the elog and object level history of the input data object. Note that the object level history is like a chain. This step converts a Seismogram into TimeSeries (or the corresponding Ensemble type), and should be considered a part of the chain.

File: python/mspasspy/algorithms/basic.py [hunk 2]
```diff
@@ -1,5 +1,55 @@
...
+        empty.kill()
+        if data.dead():
+            return empty
+        try:
+            d = bsc.ExtractComponent(data, component)
+            return d
+        except Exception as err:
+            data.elog.log_error("ExtractComponent", str(err), ErrorSeverity.Invalid)
+            return empty
+    elif isinstance(data, SeismogramEnsemble):
```
Comment (wangyinz) [inline python/mspasspy/algorithms/basic.py:38]: This has the same issue as above. Although we currently don't have a history for the ensemble type, the members of the ensemble may still container history that should be retained.

**Review by wangyinz — CHANGES_REQUESTED** [review]

File: python/mspasspy/algorithms/basic.py (hunk 1 above)
Comment (Aristoeu) [reply python/mspasspy/algorithms/basic.py:28]: I will add load_history() here to copy the ProcessingHistory of input data to the new TimeSeries object

File: python/mspasspy/algorithms/basic.py (hunk 2 above)
Comment (Aristoeu) [reply python/mspasspy/algorithms/basic.py:43]: Here I let each member in the TimeSeriesEnsemble load history of the corresponding member in the input SeismogramEnsemble. When the input is dead or an exception occurs, the return object will have elog inherited.

File: python/mspasspy/algorithms/basic.py [hunk 3]
```diff
@@ -1,5 +1,66 @@
...
+import mspasspy.ccore.algorithms.basic as bsc
+from mspasspy.ccore.seismic import (
+    TimeSeries,
+    TimeSeriesEnsemble,
+    Seismogram,
+    SeismogramEnsemble,
+)
+
+
+@mspass_func_wrapper
```
Comment (wangyinz) [inline python/mspasspy/algorithms/basic.py:14]: Please also add docstrings (as well as necessary comments) to new functions.

File: python/tests/algorithms/test_basic.py [hunk 4]
```diff
@@ -18,6 +24,37 @@
 from helper import get_live_seismogram, get_live_timeseries, get_sin_timeseries
 
 
+def test_ExtractComponent():
```
Comment (wangyinz) [inline python/tests/algorithms/test_basic.py:27]: The tests here is far from sufficient. First, you need to make sure most (if not all) the new codes are covered, and you should double check the coverage [here](https://app.codecov.io/gh/mspass-team/mspass/pull/338). Then, you need to make sure the tests also cover reasonable amount of corner cases. This is especially important to ensure exceptions are captured as expected.

File: docs/requirements.txt [hunk 5]
```diff
@@ -2,7 +2,7 @@ sphinx == 4.4.0
 sphinx-rtd-theme
 m2r2
 breathe
-nbconvert==6.5.1
+nbconvert==5.6.1
```
Comment (wangyinz) [inline docs/requirements.txt:5]: Please also resolve the version issue here.

**Review by wangyinz — CHANGES_REQUESTED** [review]
Please see the detail comments. Thanks!

**Review by wangyinz — APPROVED** [review]
